Implement psa_hash_compute and psa_hash_compare

gilles-peskine-arm opened the pull request:
A step on the way to IOTCRYPT-933

If #179 is merged first, this PR will need to be modified to add test case dependencies on `!MBEDTLS_SHA512_NO_SHA384`.

gilles-peskine-arm commented:
CI passed except for expected failures (inherited from #325): an ABI change (two external symbols have been renamed) and generated files in TLS (due to the new header).

gilles-peskine-arm commented:
CI succeeded except for `check_generated_files.sh` in TLS which is expected to fail until the crypto submodule is updated to include `crypto_compat.h` in the VS project template.

mpg commented:
Also, a conflict has developed and needs to be resolved.

gilles-peskine-arm commented:
I rebased to avoid the merge conflict. I kept the same history except for the [missing blank line between two test cases](https://github.com/ARMmbed/mbed-crypto/pull/327#discussion_r371757134).

gilles-peskine-arm commented on tests/suites/test_suite_psa_crypto.data:
```diff
@@ -766,6 +766,57 @@ hash_verify_bad_args:
+
+PSA hash compute: good, RIPEMD160
+depends_on:MBEDTLS_RIPEMD160_C
+hash_compute_verify:PSA_ALG_RIPEMD160:"616263":"8eb208f7e05d987a9b044a8e98c6b087f15a0bfc"
```
I somehow committed it in a later commit. Since I was rebasing due to the merge conflict in `check-test-cases.py`, I modified the appropriate commit to add the missing newline.

gilles-peskine-arm commented:
We don't run any tests without `MBEDTLS_MD_C`. There isn't much you can do without it! I made the changes because I prefer to be more correct, even if we can't be absolutely correct.

gilles-peskine-arm commented on tests/suites/test_suite_psa_crypto.data:
```diff
@@ -766,6 +766,57 @@ hash_verify_bad_args:
 hash_finish_bad_args:
 
+PSA hash compute: bad algorithm (unknown hash)
+depends_on:MBEDTLS_SHA256_C
```
The dependency is wrong. Probably copypasta.

gilles-peskine-arm commented on tests/suites/test_suite_psa_crypto.function:
```diff
@@ -2347,6 +2347,89 @@ exit:
+
+    PSA_ASSERT( psa_crypto_init( ) );
+
+    status = psa_hash_compute( alg, input->x, input->len,
```
ITYM `psa_hash_compare`. Indeed, I added some test cases.

mpg replied:
I think I meant `psa_hash_verify` because I was mislead by the title of this PR, the first commit message, and the name of the next test function: `hash_compute_verify`.

mpg commented on tests/suites/test_suite_psa_crypto.function:
```diff
@@ -2430,6 +2430,89 @@ exit:
+/* END_CASE */
+
+/* BEGIN_CASE */
+void hash_compute_verify( int alg_arg, data_t *input,
```
This should probably be `hash_compute_compare`. Also, the commit message could be amended.

gilles-peskine-arm replied:
Ok. I went and amended the commit that introduced this test function and reworded the commit messages.

mpg commented:
@gilles-peskine-arm  Just to notify you that #179 was just merged, so this will need a few `!MBEDTLS_SHA512_NO_SHA384` in relevant places now. Also, there's a conflict in `test_suite_psa_crypto.data` which is probably related to dependencies being updated on both sides as well.

gilles-peskine-arm commented:
Rebased again (previous iteration in https://github.com/gilles-peskine-arm/mbed-crypto/tree/psa-hash_compute-5):

* "Implement and test psa_hash_compute, psa_hash_compare": added `!MBEDTLS_SHA512_NO_SHA384` to the one new SHA-384 test case.
* "Remove obsolete dependencies on MBEDTLS_MD_C", "Add missing dependencies on MBEDTLS_MD_C": do it to the same test cases, from the updated dependencies.

yanesca commented on include/psa/crypto.h:
```diff
@@ -932,7 +932,7 @@ psa_status_t psa_hash_compare(psa_algorithm_t alg,
                               const uint8_t *input,
                               size_t input_length,
                               const uint8_t *hash,
-                              const size_t hash_length);
```
Does this mean that we need to update the API reference? This is still const in RC2...

gilles-peskine-arm replied:
Indeed, thanks for noticing that.

gilles-peskine-arm replied:
And I also fixed this in the similar function for MAC, because it's easier to fix it now than track it for later.

athoelke replied:
I'll update the spec - this doesn't actually change the interface, but it is not standard practice to use this in the API declaration.

gilles-peskine-arm replied:
@athoelke  https://github.com/ARMmbed/psa-crypto-api/pull/366

athoelke replied:
I saw that - and I've patched it with the corrected header file.

yanesca commented on tests/scripts/check-test-cases.py:
```diff
@@ -20,26 +20,30 @@
 #
 # This file is part of Mbed TLS (https://tls.mbed.org)
 
+import argparse
```
Why are the changes needed in this file/commit needed in this PR?

gilles-peskine-arm replied:
The new option to `check-test-cases` isn't needed to make `psa_hash_compute` work. But I made it and used it during the development of the PR to make it easier to see errors (they tend to get lost among the warnings).

gilles-peskine-arm commented:
CI passes except for Mbed OS jobs which is a known preexisting issue.

mpg commented:
After the latest commit, the CI is still passing except for Mbed OS tests which are a known issue independent from this PR.
